src/runner.js

```diff
@@ -28,12 +28,96 @@ const debugResponse = _debug('remote-s3:Runner:res');
+   * with its status.
+   *
+   * http://docs.aws.amazon.com/general/latest/gr/api-retries.html
+   */
   async run(opts) {
-    return this.runOnce(opts);
+    let {req, body, streamingOutput, followRedirect, currentRedirect, maxRedirects} = opts;
+    let current = 0;
+
+    currentRedirect = currentRedirect || 0;
+    maxRedirects = maxRedirects || 10;
+
+    // Since streams aren't seekable in node, we want to ensure that
+    // whenever 
+    if (body && body instanceof stream.Readable || body && typeof body.pipe === 'function') {
+      debug('WARNING: instances of stream.Readable are not retryable or redirectable');
+      current = this.maxRetries - 1;
+      followRedirect = false;
+    }
+
+    function sleepFor(n) {
+      return new Promise((resolve, reject) => {
+        let delay = Math.pow(2, n) * 100;
+        setTimeout(resolve, delay);
+      });
+    }
+
+    // Note that we've declared current earlier so that we could ensure that
+    // raw streams do not allow for retry
+    for ( ; current < this.maxRetries; current++) {
+      let result = await this.runOnce(opts);
+      if (result.statusCode >= 500 && current !== this.maxRetries - 1) {
+        let errorSummary = `${result.statusCode || '???'}: ${result.statusMessage || '???'}`;
+        debug(`RETRYABLE ERROR ${req.method} ${req.url} --> ${errorSummary}`);
+        await sleepFor(current);
+        continue;
+      } else if (followRedirect) {
+        // https://www.ietf.org/rfc/rfc2616.txt -- 10.3 Redirection 3xx
+        // These are the only 300 series codes that we're sure we can safely
+        // redirect.  Basically, anything which is not a known to be safe
+        // redirect will be returned as the 300 series response rather than
+        // following them.  If a safe to redirect 300 is missing a location
+        // header, then it will be returned as the 300 result as well
+        switch (result.statusCode) {
+          case 301:
+          case 302:
+          case 303:
+            if (req.method === 'GET' || req.method === 'HEAD') {
+              let location = result.headers.location;
+              if (location && currentRedirect < maxRedirects) {
+                return await(this.run({
+                  req: {
+                    url: location,
+                    method: req.method,
+                    headers: req.headers,
+                  },
+                  body,
+                  streamingOutput,
+                  followRedirect,
+                  maxRedirects,
+                  currentRedirect: currentRedirect + 1,
+                }));
+              } else {
+                debug(`WARNING: will only follow ${result.statusCode} with Location header`);
+              }
+            } else {
+              debug(`WARNING: will only follow ${result.statusCode} for GET or HEAD`);
+            }
+            break;
+          default:
+            debug('WARNING: will only follow a 301, 302 or 303 redirect');
```

Comment:
So reading up on S3 docs, see:
 * http://docs.aws.amazon.com/AmazonS3/latest/dev/Redirects.html
 * http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTRedirect.html
 * http://docs.aws.amazon.com/AmazonS3/latest/dev/VirtualHosting.html

We'll get redirects immediately after bucket creation in non-default region, or if we don't use the virtual style hosting `<bucket>.<region>.amazonaws.com/<key>` to access a bucket in the non-default region.

S3 will issue:
 1) 307 during DNS propagation for newly created buckets in non-default region (if using virtual style hosting)
 2) 308 if you're using the wrong end-point

I seems to me that we can avoid supporting redirects. Because we always want to use virtual style hosting.
So the few minutes after bucket creation when DNS is propagating isn't an issue.
Further more, it seems better to not follow a 308 redirect and instead consider it an error, as it means we're using the wrong end-point; which is an in-efficiency we should force ourselves to fix.

So with that in mind I propose we don't implement redirects at all. Since it's a complication we don't need.

Finally, if we really do want to support it... we should do definitely being doing `100-continue` requests before sending the payload.

Also we would have to test if it is necessary to resign the `authentication` header in the request, given that the redirected request is hitting a new hostname, and the hostname is encoded in the `Authentication` header:
http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html

In fact given that there is no way for the host to which we are redirect to know what the old hostname is, I don't see how it would be able to validate the `authentication` header unless it is regenerated..